### PR TITLE
Add Game Mode Fix

### PR DESCRIPTION
**Pull request by Vanessamae23**
Add Game Mode Fix
- Add the survival mode based on levels
- Add opponent HUD
- Add game results
- Fix render message

### CODE REVIEW

File: TowerForge/TowerForge/GameModule/Events/PlayerEvents/WaveSpawnEvent.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+struct WaveSpawnEvent: TFEvent {
+    let timestamp: TimeInterval
+    let position: CGPoint
+    let entityType: (PlayerSpawnable & TFEntity).Type
+    let player: Player
+
+    /// TODO: might need to change this to something other than throwing fatal error,
+    /// because it could cause accidental application crash, for example iterating through
+    /// a collection of TFEvents.
+    var entityId: UUID {
```
Comment (zheng-ze) [inline TowerForge/TowerForge/GameModule/Events/PlayerEvents/WaveSpawnEvent.swift:19]: Wpuld it be better to modify TFEvent to not require a entityid? Like rename the existing TFEvent to TFEntityEvent and create a new base TFEvent.

File: TowerForge/TowerForge/GameModule/GameModes/DeathMatchMode.swift [hunk 2]
```diff
@@ -11,7 +11,11 @@ class DeathMatchMode: GameMode {
     var eventManager: EventManager
     var modeName: String = "Death Match Mode"
     var modeDescription: String = "Kill as many units within the time limit"
-    var gameProps: [any GameProp] = [DeathProp(), PointProp(initialPoint: 0)]
+    var gameProps: [any GameProp] = [DeathProp(position: PositionConstants.DEATH_MATCH_POINT_OWN,
+                                               player: .ownPlayer),
+                                     DeathProp(position: PositionConstants.DEATH_MATCH_POINT_OPP,
+                                               player: .oppositePlayer),
+                                     PointProp(initialPoint: 0)]
     var timer = TimerProp(timeLength: TimeInterval(60))
```
Comment (zheng-ze) [inline TowerForge/TowerForge/GameModule/GameModes/DeathMatchMode.swift:19]: Should we increase the length of deathmatch mode? 1 minute seems a little short to be able to do anything substantial.

File: TowerForge/TowerForge/GameModule/GameModes/SurvivalGameMode.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+import Foundation
+import UIKit
+
+class SurvivalGameMode: GameMode {
+    var modeName: String = "Survival Game Mode"
+
+    var modeDescription: String = "Don't let the enemy invade your base!"
+
+    var gameProps: [any GameProp] = [PointProp(initialPoint: 0)]
```
Comment (zheng-ze) [inline TowerForge/TowerForge/GameModule/GameModes/SurvivalGameMode.swift:16]: I think you need LifeProp here as well?

**Review by zheng-ze — APPROVED** [review]
Looks good. Left a few comments to note.

File: TowerForge/TowerForge/GameModule/GameModes/SurvivalGameMode.swift (hunk 3 above)
Comment (Vanessamae23) [reply TowerForge/TowerForge/GameModule/GameModes/SurvivalGameMode.swift:16]: LifeProp is only 1 live for survival mode so only one life though.

Comment (Vanessamae23) [reply TowerForge/TowerForge/GameModule/GameModes/SurvivalGameMode.swift:16]: But maybe yea it s better to let the gamers know, will add this

**Pull request merged**